File: samples/chat/app/chat/ChatEngine.scala
package chat

import akka.stream.Materializer
import akka.stream.scaladsl.{BroadcastHub, Flow, Keep, MergeHub}
import play.api.libs.json.Format
import play.engineio.EngineIOController
import play.api.libs.functional.syntax._
import play.socketio.SocketIO

object ChatProtocol {

  case class ChatMessage(message: String)

  object ChatMessage {
    implicit val format: Format[ChatMessage] = implicitly[Format[String]]
      .inmap[ChatMessage](ChatMessage.apply, _.message)
  }

  import play.socketio.SocketIOEventCodec._

  val decoder = decoders(
    decodeJson[ChatMessage]("chat message")
  )

  val encoder = encoders {
    case _: ChatMessage => encodeJson[ChatMessage]("chat message")
  }
}

class ChatEngine(socketIO: SocketIO)(implicit mat: Materializer) {

  import ChatProtocol._

  private val chatFlow = {
    val (sink, source) = MergeHub.source[ChatMessage]
      .toMat(BroadcastHub.sink)(Keep.both).run

    Flow.fromSinkAndSourceCoupled(sink, source)
  }

  val controller: EngineIOController = socketIO.builder
    .addNamespace("/chat", decoder, encoder, chatFlow)
    .createController()
}

File: samples/java/chat/app/chat/ChatEngine.java
package chat;

import org.apache.pekko.NotUsed;
import org.apache.pekko.japi.Pair;
import org.apache.pekko.stream.Materializer;
import org.apache.pekko.stream.javadsl.*;
import play.engineio.EngineIOController;
import play.socketio.javadsl.SocketIO;
import play.socketio.javadsl.SocketIOEventCodec;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

@Singleton
public class ChatEngine implements Provider<EngineIOController> {

  private final EngineIOController controller;

  @Inject
  public ChatEngine(SocketIO socketIO, Materializer materializer) {

    // Here we define our codec. We handle chat messages in both directions as JSON strings
    SocketIOEventCodec<String, String> codec = new SocketIOEventCodec<String, String>() {
      {
        addDecoder("chat message", decodeJson(String.class));
        addEncoder("chat message", String.class, encodeJson());
      }
    };

    // We use a MergeHub to merge all the incoming chat messages from all the
    // connected users into one flow, and we feed that straight into a
    // BroadcastHub to broadcast them out again to all the connected users.
    // See http://doc.akka.io/docs/akka/2.6/scala/stream/stream-dynamic.html
    // for details on these features.
    Pair<Sink<String, NotUsed>, Source<String, NotUsed>> pair = MergeHub.of(String.class)
        .toMat(BroadcastHub.of(String.class), Keep.both()).run(materializer);

    // We couple the sink and source together so that one completes, the other
    // will to, and we use this to handle our chat
    Flow<String, String, NotUsed> chatFlow = Flow.fromSinkAndSourceCoupled(pair.first(), pair.second());

    // Here we create an EngineIOController to handle requests for our chat
    // system, and we add the chat flow under the "/chat" namespace.
    controller = socketIO.createBuilder()
        .addNamespace("/chat", codec, chatFlow)
        .createController();
  }

  public EngineIOController get() {
    return controller;
  }
}